src/disco/mux/fd_mux.h

```diff
@@ -51,6 +92,152 @@
+
+   The ctx is a user-provided context object from when the mux tile was
+   initialized. */
+typedef void (fd_mux_before_poll_fn)( void * ctx );
+
+typedef void (fd_mux_after_poll_fn)( void *             ctx,
+                                     fd_mux_context_t * mux );
+
+typedef void (fd_mux_before_frag_fn)( void * ctx,
+                                      ulong  in_idx,
+                                      ulong  sig,
+                                      ulong  seq,
+                                      int *  opt_filter );
+
+/* fd_mux_during_frag_fn is called after the mux has received a new frag
+   from an in, but before the mux has checked that it was overrun.  This
+   callback is not invoked if the mux is backpressured, as it would not
+   try and read a frag from an in in the first place (instead, leaving
+   it on the in mcache to backpressure the upstream producer).  in_idx
+   will be the index of the in that the frag was received from.
+
+   If the producer of the frags is respecting flow control, it is safe
+   to read frag data in any of the callbacks, but it is suggested to
+   copy or read frag data within this callback, as if the producer does
+   not respect flow control, the frag may be torn or corrupt due to an
+   overrun by the reader.  If the frag being read from has been
+   overwritten while this callback is running, the frag will be ignored
+   and the mux will not call the process function.  Instead it will
+   recover from the overrun and continue with new frags.
+
+   This function cannot fail.  If opt_filter is set to non-zero, it
+   means the frag should be filtered and not passed on to downstream
+   consumers of the mux.
+
+   The ctx is a user-provided context object from when the mux tile was
+   initialized.
+
+   sig, chunk, and sz are the respective fields from the mcache fragment
+   that was received.  If the producer is not respecting flow control,
+   these may be corrupt or torn and should not be trusted. */
+
+typedef void (fd_mux_during_frag_fn)( void * ctx,
+                                      ulong  in_idx,
+                                      ulong  sig,
+                                      ulong  chunk,
+                                      ulong  sz,
+                                      int *  opt_filter );
+
+/* fd_mux_after_frag_fn is called immediately after the
+   fd_mux_during_frag_fn, along with an additional check that the reader
+   was not overrun while handling the frag.  If the reader was overrun,
+   the frag is abandoned and this function is not called.  This callback
+   is not invoked if the mux is backpressured, as it would not read a
+   frag in the first place.  It is also not invoked if
+   fd_mux_during_frag returns non-zero, indicating to filter the frag.
+
+   You should not read the frag data directly here, as it might still
+   get overrun, instead it should be copied out of the frag during the
+   read callback if needed later.
```

Comment:
The name `read callback` made me start thinking about the names of these callbacks. What do you think about naming `fd_mux_during_frag_fn` something like speculative read. Then `fd_mux_after_frag_fn` -> read confirmed or something like that?